src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs

```diff
@@ -384,38 +392,51 @@ protected unsafe int UnsafeGetUTF8Chars(char* chars, int charCount, byte[] buffe
                 fixed (byte* _bytes = &buffer[offset])
                 {
                     byte* bytes = _bytes;
-                    byte* bytesMax = &bytes[buffer.Length - offset];
                     char* charsMax = &chars[charCount];
 
-                    while (true)
+                    if (Sse41.IsSupported && charCount >= Vector128<short>.Count)
```

Comment:
What is the purpose of this UnsafeGetUTF8Chars method?  Why isn't Encoding.UTF8.GetBytes just used instead?  It's already highly-optimized.

Reply:
I think it is to avoid the overhead of the method call since many element names are relatively small 10-30? character and all ascii.

Even the original simple loop was better for up to 20 something characters and both the proposed sse and avx version (se link in question 1. In summary)  are more than twice as fast than system encoding for length [8, ~100 ( se [full results](https://github.com/Daniel-Svensson/ClrExperiments/blob/master/BinaryXmlBenchmarks/ConsoleApp1/BenchmarkDotNet.Artifacts/results_amd_5800X/ConsoleApp1.Utf8Benchmarks-report-3.md)


It is only at the maximum lenght of 170 for this call path where System encoding comes somewhat near the sse version and still quite far from avx alternative.

Reply:
Can you share the benchmark you're using?  I see ones in the same repro as the linked "full results", but I don't see code doing `Encoding.UTF8.GetBytes`; the closest it gets is `s_encoding.GetBytes`, which is not the same.

Reply:
> Can you share the benchmark you're using? I see ones in the same repro as the linked "full results", but I don't see code doing Encoding.UTF8.GetBytes; the closest it gets is s_encoding.GetBytes, which is not the sam

 
Yes the [Benchmkared call is](https://github.com/Daniel-Svensson/ClrExperiments/blob/97e6caa6a77c8c64f9d231b2647376e74043efe1/BinaryXmlBenchmarks/ConsoleApp1/Utf8Benchmarks.cs#L80) `s_encoding.GetBytes`  because that is what it original code is using and I cannot rule out that changing the call to `Encoding.UTF8.GetBytes` could change the behaviour somehow since the  encoding since `new UTF8Encoding(false, true);` is used instead of the default  constructor.

I think the closes thing I did benchmark was  to call [System.Text.Unicode.Utf8.FromUtf16](https://github.com/Daniel-Svensson/ClrExperiments/blob/97e6caa6a77c8c64f9d231b2647376e74043efe1/BinaryXmlBenchmarks/ConsoleApp1/Utf8Benchmarks.cs#L90) but that was not faster than calling ` s_UTF8Encoding.GetBytes` so I only run it once.

Link: https://github.com/Daniel-Svensson/ClrExperiments/blob/97e6caa6a77c8c64f9d231b2647376e74043efe1/BinaryXmlBenchmarks/ConsoleApp1/Utf8Benchmarks.cs

Reply:
Calling `Encoding.UTF8.GetBytes(...)` is guaranteed to have the same output as `new UTF8Encoding(false).GetBytes(...)`. The only difference between the two is the output of the `UTF8Encoding.Preamble` property, which is not queried in these code paths. Like Steve said, you should see a perf increase if you call `Encoding.UTF8.GetBytes(...)` directly due to the JIT's ability to devirtualize the method call.

Reply:
The code had been changing from using UTF8Encoding with `throwOnInvalidBytes: false` to `throwOnInvalidBytes: true`
So it seems safe to use Encoding.UTF8 instead now, but I will split out making changes to general UTF8 encoding from this PR in order to avoid causing any performance regression causing the whole PR to be reverted